Auto Takeoff

NixonCTChan opened the pull request:
- This draft includes the previous landing stuff
- Added takeoff states, functions, and tests
- Added reset variables state and the reset function
- Added passby structure to waypoint manager (some dependency issues come up if I include it in the landingTakeoff file, should I keep it somewhere else?)
- Added passby functionality to attitude manager

Pop0097 commented on Autopilot/Test/Src/LandingTakeoffManager/Test_Takeoff.cpp:
```diff
@@ -0,0 +1,60 @@
+    EXPECT_NEAR(desiredTakeoffPoint.latitude, 1.0006353018, 0.000001);
+    EXPECT_NEAR(desiredTakeoffPoint.longitude, 2.0006352019, 0.000001);
+    EXPECT_NEAR(desiredTakeoffPoint.altitude, 3, 0.000001);
+}
```
Need newline

NixonCTChan replied:
gotcha

Pop0097 commented on Autopilot/AttitudeManager/Src/attitudeStateClasses.cpp:
```diff
@@ -92,11 +92,30 @@ void PIDloopMode::execute(attitudeManager* attitudeMgr)
     PMCommands pathManagerOutput;
     PMError_t pmError = PM_GetCommands(&pathManagerOutput);
 
+    //executes PID's to acheive desired roll, pitch angle
     _PidOutput.rollPercent = _rollPid.execute(PMInstructions->roll, SFOutput->IMUroll, SFOutput->IMUrollrate);
     _PidOutput.pitchPercent = _pitchPid.execute(PMInstructions->pitch, SFOutput->IMUpitch, SFOutput->IMUpitchrate);
     _PidOutput.rudderPercent = pathManagerOutput.rudderPercent;
     _PidOutput.throttlePercent = pathManagerOutput.throttlePercent;
 
+    //if manual control is needed, use loaded in percents instead!
+    if(pathManagerOutput.passbyData.pitchPassby)
+    {
+        _PidOutput.pitchPercent = pathManagerOutput.passbyData.pitchPercent;
+    } 
+    if(pathManagerOutput.passbyData.rollPassby)
+    {
+        _PidOutput.rollPercent = pathManagerOutput.passbyData.rollPercent;
+    } 
+    if(pathManagerOutput.passbyData.rudderPassby)
+    {
+        _PidOutput.rudderPercent = pathManagerOutput.passbyData.rudderPercent;
+    } 
+    if(pathManagerOutput.passbyData.throttlePassby)
+    {
+        _PidOutput.throttlePercent = pathManagerOutput.passbyData.throttlePercent;
+    } 
+
```
imo it would be better to do if else statements here. No need to execute the PID if we are just going to override the value anyway yknow?

NixonCTChan replied:
true, fixed!

Pop0097 commented on Autopilot/PathManager/Inc/landingTakeoffManager.hpp:
```diff
@@ -0,0 +1,102 @@
+    LANDING SECTION
+************************/
+
+constexpr static double DISTANCE_OF_LANDING {10}; //in meters
+constexpr static double METERS_PER_DEG_LAT {111320.0};
+constexpr static double ANGLE_OF_LANDING {5}; //in degrees
+constexpr static int STALL_SPEED_NO_PACKAGE {2}; //in m/s
+constexpr static int STALL_SPEED_WITH_PACKAGE {4}; //in m/s
+constexpr static double PI {3.14159265358979323846};
+constexpr static int FLARE_ALTITUDE {2}; //in meters
+constexpr static double DECRAB_ALTITUDE {0.7}; //in meters
+constexpr static double TOUCHDOWN_ALTITUDE {0.05}; //in meters
```
Curious, why are you not using `#define` here?

NixonCTChan replied:
I was talking to Anthony about this, he recommended constexpr to define constants at compile time. Is there any advantage to use define instead?

Pop0097 replied:
https://stackoverflow.com/questions/42388077/constexpr-vs-macros/42388188

Covers the differences well. I agree we should stick to constexpr :))

Pop0097 commented on Autopilot/PathManager/Inc/landingTakeoffManager.hpp:
```diff
@@ -0,0 +1,102 @@
+    double landingDirection;
+};
+
+class LandingManager : public WaypointManager{
+    
+    public:
+        /*
+            This function takes the horizontal position of the aircraft and determines the desired altitude
+            Projects the position to the vertical plane relative to the aimingPoint, intersectionPoint, and stoppingPoint
+            Takes the one of the horizontal coordinates, plug it into the equation of the slope, get an altitude
+        */
+        static double changingAltitude(_WaypointManager_Data_In input, _PathData aimingPoint, _PathData intersectionPoint, _PathData stoppingPoint);
+
+        /*
+            This function sets the desired altitude so that the throttle turns off
+        */
+        static double throttleOff(void);
+
+        /*
+            This function takes in the current magnitude of windspeed, and returns an approach speed the aircraft should follow
+        */
+        static double approachSpeed(double windSpeed, bool ifPackage);
+
+        /* 
+            This function takes the stall speed and multiplies it by a constant to return a slow flight speed
+        */
+        static double slowFlightSpeed(bool ifPackage);
+
+        /*
+            This function takes the coordinates of the landing spots and direction of landing
+            Creates stopping points, aiming point, and intersection point
+        */
+        static _LandingPath createSlopeWaypoints(Telemetry_PIGO_t input, double currentAltitude);
+
+};
```
I would strongly advise against using inheritance here. Instead, make a private class member that is an instance of the waypoint manager.

In good OOP design, the derived class should be a specialization of the parent class (the design principle is called Liskov's Principle of Substitutability). I don't think this matches that case... @Fberbari what do you think?

NixonCTChan replied:
Yeah now that I think about it again, thats true. I'll remove the inheritance completely. Since its just a class of functions, there doesnt need to be any connections between waypoint manager and landing takeoff stuff.

Pop0097 commented on Autopilot/PathManager/Inc/landingTakeoffManager.hpp:
```diff
@@ -0,0 +1,102 @@
+
+struct _LandingInitialData
+{
+    double stoppingLatitude;
+    double stoppingLongitude;
```
should be `long double` to keep consistency with other structs

NixonCTChan replied:
gotcha

Pop0097 commented on Autopilot/PathManager/Inc/landingTakeoffManager.hpp:
```diff
@@ -0,0 +1,102 @@
+        */
+        static double changingAltitude(_WaypointManager_Data_In input, _PathData aimingPoint, _PathData intersectionPoint, _PathData stoppingPoint);
+
+        /*
+            This function sets the desired altitude so that the throttle turns off
+        */
+        static double throttleOff(void);
```
How does it set the desired altitude if it doesn't take in a parameter?

NixonCTChan replied:
Yeah forgot to remove these functions, their functionality is being replaced with the passbyControl structure.

Pop0097 commented on Autopilot/PathManager/Inc/landingTakeoffManager.hpp:
```diff
@@ -0,0 +1,102 @@
+    TAKEOFF SECTION
+************************/
+
+constexpr static double EXIT_TAKEOFF_ALTITUDE {50}; //in meters, difference in altitude
+constexpr static double ROTATION_SPEED_NO_PACKAGE {4}; //in meters per second
+constexpr static double ROTATION_SPEED_WITH_PACKAGE {6}; //in meters per second
+constexpr static double CLIMB_SPEED_NO_PACKAGE {5}; // in meters per second
+constexpr static double CLIMB_SPEED_WITH_PACKAGE {7}; // in meters per second
+constexpr static double DISTANCE_OF_TAKEOFF {100}; // in meters, just used to approximate the horizontal location
```
Why are we using `constexpr` instead of preprocessors?

NixonCTChan replied:
Answered above

Pop0097 commented on Autopilot/PathManager/Inc/landingTakeoffManager.hpp:
```diff
@@ -0,0 +1,102 @@
+            This waypoint does not have a vertical component, as max throttle is kept throughout the climb
+        */
+        static _PathData createTakeoffWaypoint(double currentLatitude, double currentLongitude, double currentAltitude, double takeoffDirection);
+};
```
Same point as before. But now that I look at it, you may need a constructor to initialize the waypoint manager thinking about it...

Pop0097 commented on Autopilot/PathManager/Inc/pathDatatypes.hpp:
```diff
@@ -0,0 +1,44 @@
+
+};
+
+#endif
```
new line

Pop0097 commented on Autopilot/AttitudeManager/Inc/GetFromPathManager.hpp:
```diff
@@ -2,14 +2,15 @@
 	float roll, pitch;	// commanded orientation (radians)
 	float rudderPercent;
 	float throttlePercent;
+	_PassbyControl passbyData;
```
LandingTakeoffManasger.hpp doesn't have a struct called `_PassbyControl`... where is this struct coming from?

Pop0097 commented on Autopilot/PathManager/Inc/waypointManager.hpp:
```diff
@@ -18,7 +18,27 @@ struct _WaypointManager_Data_In {
+* Structure stores manual control information
+* If a certain variable needs to be controlled, set the percent to the desired value and its respective boolean to true
+*/
+struct _PassbyControl{
+    double rollPercent;
+    bool rollPassby;
+
+    double rudderPercent;
+    bool rudderPassby;
+
+    double pitchPercent;
+    bool pitchPassby;
+
+    double throttlePercent;
+    bool throttlePassby;
```
Would it be better to put this struct in `pathDatatypes.hpp`. I see no point putting it here if waypoint manager will do nothing with it

NixonCTChan replied:
yup will do

Pop0097 commented on Autopilot/PathManager/Inc/pathStateClasses.hpp:
```diff
@@ -61,12 +64,59 @@ class sensorFusion : public pathManagerState
         void execute(pathManager* pathMgr);
         void exit(pathManager* pathMgr) {(void) pathMgr;}
         static pathManagerState& getInstance();
+        static _WaypointManager_Data_In sensorInput; //must be filled out
```
Ideally parameters should not be made public

NixonCTChan replied:
some of these variables may need to be accessed and modified by other states. Having getters and setters for each variable would get very cluttered, which is why I'm making them public. Do you want to do it with getters and setters?

Pop0097 replied:
I personally would rather we use getters and setters. @Fberbari what do you think?

Pop0097 commented on Autopilot/PathManager/Inc/pathStateClasses.hpp:
```diff
@@ -61,12 +64,59 @@ class sensorFusion : public pathManagerState
+        resetVariables() {}
+        resetVariables(const resetVariables& other);
+        resetVariables& operator =(const resetVariables& other);
+        void resetPassby(_PassbyControl* controlDetails){controlDetails->rollPassby = controlDetails->pitchPassby = controlDetails->throttlePassby = controlDetails->rudderPassby = false;}
```
I've never seen this type of syntax before... what's it called?

Pop0097 replied:
Also if there is a more readable way to do this, that would be best (also it may be worth defining the function in pathStateClasses.cpp

NixonCTChan replied:
http://courses.cms.caltech.edu/cs11/material/cpp/donnie/cpp-ops.html
refer to assignment operator chaining. I would be able to do it another way, just wanted to see if you guys were ok with chaining operators like this.

Pop0097 replied:
Damn, I didn't know we could do this. Thanks for sharing! Maybe put a little comment linking the website and saying you're doing operating chaining?

NixonCTChan replied:
Yup, I'll put it down.

Pop0097 commented on Autopilot/PathManager/Inc/pathStateClasses.hpp:
```diff
@@ -61,12 +64,59 @@ class sensorFusion : public pathManagerState
+        static _PathData * pathArray[1];
+        static _PathData * currentLocation;
+        static _WaypointStatus waypointStatus;
+        static _PathData takeoffPoint;
```
Data parameters usually should always be made private.

Pop0097 commented on Autopilot/PathManager/Inc/landingTakeoffManager.hpp:
```diff
@@ -0,0 +1,102 @@
+#pragma once
+
```
Curious, why are we putting both the landing and takeoff manager In the same file? Would it make sense to break them into separate files?

NixonCTChan replied:
some of their functions are really similar, and it provides us one file where we can tune the important speeds. Ex. approach speed, climb speed. I don't see why we need to separate them into two files

Pop0097 commented on Autopilot/PathManager/Src/landingTakeoffManager.cpp:
```diff
@@ -0,0 +1,184 @@
+#include "waypointManager.hpp"
+#include "landingTakeoffManager.hpp"
+#include <math.h>
+#include "vectorClass.hpp"
```
Don't need to include waypointManager and vectorClass as they are already included in landingTakeoffManager

Pop0097 commented on Autopilot/PathManager/Src/landingTakeoffManager.cpp:
```diff
@@ -0,0 +1,184 @@
+    Vector3D vectorAI(iPoint.x - aPoint.x, iPoint.y - aPoint.y, iPoint.z - aPoint.z);
+    Vector3D vectorAS(sPoint.x - aPoint.x, sPoint.y - aPoint.y, sPoint.z - aPoint.z);
+    Vector3D vectorAC(cPoint.x - aPoint.x, cPoint.y - aPoint.y, cPoint.z - aPoint.z);
+    Vector3D normal;
```
the variable `normal` is blue here... that kinda worries me because git is interpreting it as something. Maybe change the name to normalVector?

NixonCTChan replied:
had that issue before, it doesn't actually do anything but sure, I'll change it to normalVector.

Pop0097 commented on Autopilot/PathManager/Src/landingTakeoffManager.cpp:
```diff
@@ -0,0 +1,184 @@
+double LandingManager::slowFlightSpeed(bool ifPackage)
+{
+    //slow flight speed calculation for both package and no package scenarios
+    if(ifPackage)
+    {
+        return STALL_SPEED_WITH_PACKAGE + 2;
+    }
+    else 
+    {
+        return STALL_SPEED_NO_PACKAGE + 2;
+    }
```
would it make sense to make it slightly faster (to account for some more error)?

NixonCTChan replied:
these values will need to be tuned when we're simulating the landing takeoff stuff, they're not set in stone.

Pop0097 commented on Autopilot/PathManager/Src/landingTakeoffManager.cpp:
```diff
@@ -0,0 +1,184 @@
+#include "waypointManager.hpp"
```
Yeah, looking at your implementations, these classes should not inherit WaypointManager. If you could instead create an instance of waypoint manager in the classes that would be a lot better deign-wise

Pop0097 commented on Autopilot/PathManager/Src/pathStateClasses.cpp:
```diff
@@ -55,14 +55,11 @@ pathManagerState& getSensorData::getInstance()
-    {
-        pathMgr -> setState(cruisingState::getInstance());
-    }
+    pathMgr->setState(sensorFusion::getInstance());
```
Keep the else{} statement here for consistency please

Pop0097 commented on Autopilot/PathManager/Src/pathStateClasses.cpp:
```diff
@@ -120,4 +164,287 @@ pathManagerState& fatalFailureMode::getInstance()
+        pathArray[0] = landingPath.initialize_waypoint(path.intersectionPoint.longitude, path.intersectionPoint.latitude, path.intersectionPoint.altitude, PATH_FOLLOW);
+        pathArray[1] = landingPath.initialize_waypoint(path.aimingPoint.longitude, path.aimingPoint.latitude, path.aimingPoint.altitude, PATH_FOLLOW);
+        pathArray[2] = landingPath.initialize_waypoint(path.stoppingPoint.longitude, path.stoppingPoint.latitude, path.stoppingPoint.altitude, PATH_FOLLOW);
+        currentLocation = landingPath.initialize_waypoint(sensorFusion::sensorInput.longitude, sensorFusion::sensorInput.latitude, sensorFusion::sensorInput.altitude, HOLD_WAYPOINT); //fill in with sensor fusion data
```
If you are defining a hold waypoint, you should call this method:

`_PathData* WaypointManager::initialize_waypoint(long double longitude, long double latitude, int altitude, _WaypointOutputType waypointType, float turnRadius)`

instead

Pop0097 commented on Autopilot/PathManager/Src/pathStateClasses.cpp:
```diff
@@ -120,4 +164,287 @@ pathManagerState& fatalFailureMode::getInstance()
+
+void landingTransitionStage::execute(pathManager* pathMgr)
+{
+    if(!pathMgr -> madeLandingPoints)
```
Styling, but don't put spaces around `->`

Pop0097 commented on Autopilot/PathManager/Src/pathStateClasses.cpp:
```diff
@@ -1,6 +1,5 @@
 #include "pathStateClasses.hpp"
-
-bool isError; 
+#include "landingTakeoffManager.hpp"
```
Don't need this

Pop0097 commented on Autopilot/PathManager/Src/pathStateClasses.cpp:
```diff
@@ -120,4 +164,287 @@ pathManagerState& fatalFailureMode::getInstance()
+    waypointStatus = landingPath.get_next_directions(sensorFusion::sensorInput, &cruisingState::_outputdata);
+
+    //calculating the difference in heading to detect if finished turning (2 differences in heading possible)
+    differenceInHeading1 = getFromTelemetry::telemetryInput.stoppingDirectionHeading - sensorFusion::sensorInput.track;
+    differenceInHeading2 = sensorFusion::sensorInput.track - getFromTelemetry::telemetryInput.stoppingDirectionHeading;
```
I don't like how we are just accessing variables from other classes like this. I would rather you made getter functions since it will prevent people (in the future) from actually changing the values.

NixonCTChan replied:
Answered above

Pop0097 commented on Autopilot/PathManager/Src/pathStateClasses.cpp:
```diff
@@ -120,4 +164,287 @@ pathManagerState& fatalFailureMode::getInstance()
+    if(differenceInHeading2 < 0){differenceInHeading2 += 360;}
+
+    //if the smaller heading is less than 5 degrees, set stage to slope
+    if(differenceInHeading1<differenceInHeading2)
```
space around `<` and other comparative operators

Pop0097 commented on Autopilot/PathManager/Src/pathStateClasses.cpp:
```diff
@@ -120,4 +164,287 @@ pathManagerState& fatalFailureMode::getInstance()
+    if(differenceInHeading2 < 0){differenceInHeading2 += 360;}
+
+    //if the smaller heading is less than 5 degrees, set stage to slope
+    if(differenceInHeading1<differenceInHeading2)
+    {
+        if(fabs(differenceInHeading1) <= 5)
```
Suggestion: you can use the && operator to get rid of the nested if statements

Pop0097 commented on Autopilot/PathManager/Src/pathStateClasses.cpp:
```diff
@@ -120,4 +164,287 @@ pathManagerState& fatalFailureMode::getInstance()
+    if(!pathMgr -> madeLandingPoints)
+    {
+        //requires data structure that dhruv wants to use 
+        landingPath = WaypointManager(0.0,0.0);
+        path = LandingManager::createSlopeWaypoints(getFromTelemetry::telemetryInput, sensorFusion::sensorInput.altitude);
```
if you are creating a new waypoint manager you can actually make a constructor for LandingManager. So getting rid of the inheritance should not be too complex

NixonCTChan replied:
There doesn't need to be a constructor, landing and takeoff classes are just a class of functions.

Pop0097 commented on Autopilot/PathManager/Src/pathStateClasses.cpp:
```diff
@@ -120,4 +164,287 @@ pathManagerState& fatalFailureMode::getInstance()
+    else
+    {
+        //align heading with landing direction
+        cruisingState::_outputdata.desiredHeading = getFromTelemetry::telemetryInput.stoppingDirectionHeading;
+        //retrieving desired slow flight speed
+        cruisingState::_outputdata.desiredSpeed = LandingManager::slowFlightSpeed(getFromTelemetry::telemetryInput.packageOnBoard);
```
Why are you accessing crusingState parameters?

NixonCTChan replied:
Since coordinated turns would need to be pulling instructions from a wide variety of states, I was thinking to modify one variable for all the control states. Now that I think about it, I think it's better for all states to modify an input variable in coordinate turns instead of cruising. What do you think?

Pop0097 reviewed (commented):
Looking good so far!

Most of my comments have to do with design and styling.

Pop0097 commented on Autopilot/PathManager/Inc/waypointManager.hpp:
```diff
@@ -11,6 +11,7 @@
 #include <stdlib.h>
 #include <math.h>
 #include <cstdint>
+#include "pathDatatypes.hpp"
```
Will waypoint manager be using any of the structs in here? If not, then no need to include it

NixonCTChan replied:
yeah, the passby control is in the waypoint manager out struct. Should I make it separate?

Pop0097 commented on Autopilot/PathManager/Src/pathStateClasses.cpp:
```diff
@@ -202,13 +204,11 @@ void landingTransitionStage::execute(pathManager* pathMgr)
-        }
+        pathMgr->stage = SLOPE;
     }
     else
```
use `else if ()` here instead.

Pop0097 reviewed (commented):
Nice improvements. 

Some comments still have to be addressed (like public class members and all, but overall fix seems good so far).

NixonCTChan commented:
Changes made:
- added landing and takeoff state transition tests
- added enum in CT to determine what instructions to pull from
- added functionality to telemetry input for beginLanding and beginTakeoff
- created landing and takeoff I/O structs, and the translation functions for those
- created variable definitions at the top of pathStateClasses.cpp to fix undefined reference error
- integrated cruising state I/O functions into the enums for both resetVariables and CT

Pop0097 commented on Autopilot/PathManager/Inc/pathDatatypes.hpp:
```diff
@@ -0,0 +1,75 @@
+    bool beginLanding;
+    bool beginTakeoff;
+    float stoppingDirectionHeading;
+    float stoppingLongitude;
+    float stoppingLatitude;
+    float stoppingAltitude;
```
I'm assuming all of this will be inside of the stopping point gps struct?

NixonCTChan replied:
do you mean if I'm gonna create a new struct to hold all these in?

Fberbari replied:
I'd be ok with either creating a new struct or not. No big deal either way.

Pop0097 commented on Autopilot/PathManager/Inc/landingTakeoffManager.hpp:
```diff
@@ -0,0 +1,132 @@
+
+enum _FlightStage{ROLL=0, CLIMB, CRUISING, TRANSITION, SLOPE, FLARE, DECRAB, TOUCHDOWN};
+
+struct _ManualControl
+{
+    double rollPercent;
+    bool rollManual;
+
+    double rudderPercent;
+    bool rudderManual;
+
+    double elevatorPercent;
+    bool elevatorManual;
+
+    double throttlePercent;
+    bool throttleManual;
+};
```
This struct is a carbon copy of _PassbyControl and I couldn't seem to find a place where it is being used. Would it be worth to delete it?

NixonCTChan replied:
oops sorry, this leaked through when solving the merge conflicts

Pop0097 commented on Autopilot/PathManager/Inc/landingTakeoffManager.hpp:
```diff
@@ -0,0 +1,132 @@
+#pragma once
+
+
+#include "waypointManager.hpp"
+#include "vectorClass.hpp"
+#include "AutoSteer.hpp"
+#include "SensorFusion.hpp"
```
I would #inlude pathDatatypes.hpp since you are referencing _PassbyControl here :))

Pop0097 commented on Autopilot/PathManager/Inc/landingTakeoffManager.hpp:
```diff
@@ -0,0 +1,132 @@
+            This function takes the horizontal position of the aircraft and determines the desired altitude
+            Projects the position to the vertical plane relative to the aimingPoint, intersectionPoint, and stoppingPoint
+            Takes the one of the horizontal coordinates, plug it into the equation of the slope, get an altitude
+        */
```
Love the comments. Little room for improvement: For comments on methods, this structure is usually a bit better:

```
        /**
          *  This function takes the horizontal position of the aircraft and determines the desired altitude
          * Projects the position to the vertical plane relative to the aimingPoint, intersectionPoint, and stoppingPoint
          *  Takes the one of the horizontal coordinates, plug it into the equation of the slope, get an altitude
          * 
          * @param WRITE ONE OF THESE STATEMENTS FOR EACH PARAMETER
          * 
          * @return WHAT IS BEING RETURNED? 
        */
```

Pop0097 commented on Autopilot/PathManager/Src/pathStateClasses.cpp:
```diff
@@ -90,16 +231,79 @@ pathManagerState& cruisingState::getInstance()
+        case CLIMB:
+            LandingManager::translateLTSFCommandsToCoordTurns(takeoffClimbStage::getControlOutput(), sensFusionOutput, imudata, &turnInput, &altAirspeedInput);
+            break;
+        case TRANSITION:
+            LandingManager::translateLTSFCommandsToCoordTurns(landingTransitionStage::getControlOutput(), sensFusionOutput, imudata, &turnInput, &altAirspeedInput);
+            break;
+        case SLOPE:
+            LandingManager::translateLTSFCommandsToCoordTurns(landingSlopeStage::getControlOutput(), sensFusionOutput, imudata, &turnInput, &altAirspeedInput);
+            break;
+        case FLARE:
+            LandingManager::translateLTSFCommandsToCoordTurns(landingFlareStage::getControlOutput(), sensFusionOutput, imudata, &turnInput, &altAirspeedInput);
+            break;
+        case DECRAB:
+            LandingManager::translateLTSFCommandsToCoordTurns(landingDecrabStage::getControlOutput(), sensFusionOutput, imudata, &turnInput, &altAirspeedInput);
+            break;
+        case TOUCHDOWN:
+            LandingManager::translateLTSFCommandsToCoordTurns(landingTouchdownStage::getControlOutput(), sensFusionOutput, imudata, &turnInput, &altAirspeedInput);
+            break;
+        case CRUISING:
+            turnInput.currentHeadingTrack = sensFusionOutput->track; // Gets track;
+            turnInput.desiredHeadingTrack = waypointOutput->desiredTrack;
+            turnInput.accY = imudata->accy;
+            
+            // Set up the compute altitude and airspeed function input
+            altAirspeedInput.currentAltitude = sensFusionOutput->altitude;
+            altAirspeedInput.desiredAltitude = waypointOutput->desiredAltitude;
+            altAirspeedInput.currentAirspeed = sensFusionOutput->airspeed;
+            altAirspeedInput.desiredAirspeed = waypointOutput->desiredAirspeed;   
+            break;
+        default:
+            //should default to cruising
+        
+            turnInput.currentHeadingTrack = sensFusionOutput->track; // Gets track;
+            turnInput.desiredHeadingTrack = waypointOutput->desiredTrack;
+            turnInput.accY = imudata->accy;
+            
+            // Set up the compute altitude and airspeed function input
+            altAirspeedInput.currentAltitude = sensFusionOutput->altitude;
+            altAirspeedInput.desiredAltitude = waypointOutput->desiredAltitude;
+            altAirspeedInput.currentAirspeed = sensFusionOutput->airspeed;
```
Would it make more sense to use if statements here? Just wondering since half of the conditions have the same body

NixonCTChan replied:
these statements are actually different since they pull commands from each of the different states. If I were to use if statements, there would be just as many cases.

Pop0097 replied:
Fair enough. Thanks for clearing it up!

Pop0097 commented on Autopilot/PathManager/Src/pathStateClasses.cpp:
```diff
@@ -1,11 +1,87 @@
+// Landing and Takeoff Variables (some stages DO NOT need waypoint inputs, it's not a mistake that some are missing)
+_LandingTakeoffInput takeoffRollStage::input;
+_LandingTakeoffOutput takeoffRollStage::output;
+_WaypointManager_Data_Out takeoffRollStage::waypointOutput;
+_PathData * takeoffRollStage::currentLocation;
```
You are using the identifiers "waypointOutput", etc. a lot. Maybe add the names of the states they belong to? That way it'll be a bit more obvious to a reader

This should especially be done to public members

NixonCTChan replied:
most of these are just placeholders for waypoint data that gets outputted, which are then immediately translated to landing and takeoff commands. They are all private members. I'll put a comment that explains it better

Pop0097 commented on Autopilot/PathManager/Src/pathStateClasses.cpp:
```diff
@@ -120,4 +324,360 @@ pathManagerState& fatalFailureMode::getInstance()
+    else if(fabs(differenceInHeading2) <= 5)
+    {
+        pathMgr->stage = SLOPE;
+    }
```
You can use the || operand to make this one if statement instead of two

Pop0097 commented on Autopilot/PathManager/Src/pathStateClasses.cpp:
```diff
@@ -120,4 +324,360 @@ pathManagerState& fatalFailureMode::getInstance()
+
+    //calculating the difference in heading to detect if finished turning (2 differences possible)
+    differenceInHeading1 = input.telemetryData.stoppingDirectionHeading - input.sensorOutput.track;
+    differenceInHeading2 = input.sensorOutput.track - input.telemetryData.stoppingDirectionHeading;
```
Do we need to make these static members of the state? by the looks of it we aren't referring to past values that they held and other states aren't using them.

Pop0097 commented on Autopilot/PathManager/Src/pathStateClasses.cpp:
```diff
@@ -120,4 +324,360 @@ pathManagerState& fatalFailureMode::getInstance()
+        pathArray[2] = landingPath.initialize_waypoint(path.stoppingPoint.longitude, path.stoppingPoint.latitude, path.stoppingPoint.altitude, PATH_FOLLOW);
+        currentLocation = landingPath.initialize_waypoint(input.sensorOutput.longitude, input.sensorOutput.latitude, input.sensorOutput.altitude, HOLD_WAYPOINT, 20); //fill in with sensor fusion data
+        //initializing flight path
+        waypointStatus = landingPath.initialize_flight_path(pathArray, 3, currentLocation);
```
Amazing job using the API! I hope it was easy to use!

Pop0097 commented on Autopilot/PathManager/Src/pathStateClasses.cpp:
```diff
@@ -120,4 +324,360 @@ pathManagerState& fatalFailureMode::getInstance()
+        //retrieving desired altitude for slope state and setting it 
+        output.desiredAltitude = LandingManager::changingAltitude(input.sensorOutput,landingTransitionStage::path.aimingPoint, landingTransitionStage::path.intersectionPoint, landingTransitionStage::path.stoppingPoint);
+        //retrieving desired speed for approach speed and setting it
+        output.desiredAirspeed = LandingManager::approachSpeed(pathMgr->isPackage);
```
Adding some spacing like you did in landingFlareStage::execute() will make this more readable!

Pop0097 commented on Autopilot/PathManager/Src/pathStateClasses.cpp:
```diff
@@ -120,4 +324,360 @@ pathManagerState& fatalFailureMode::getInstance()
+
+    if(landingTransitionStage::waypointStatus == INVALID_PARAMETERS)
+    {
+        pathMgr -> isError = true;
```
Please remove spaces around `->`. There are a few places where you did this

Pop0097 commented on Autopilot/PathManager/Src/pathStateClasses.cpp:
```diff
@@ -120,4 +324,360 @@ pathManagerState& fatalFailureMode::getInstance()
+        output.desiredAirspeed = LandingManager::slowFlightSpeed(pathMgr->isPackage);
+    }   
+
+    if(landingTransitionStage::waypointStatus == INVALID_PARAMETERS)
```
Probs would use waypointStatus != WAYPOINT_SUCCESS instead!

Pop0097 replied:
(There are a few error codes, but only one success code)

Pop0097 commented on Autopilot/PathManager/Src/pathStateClasses.cpp:
```diff
@@ -120,4 +324,360 @@ pathManagerState& fatalFailureMode::getInstance()
+        //align heading with landing direction
+        output.desiredHeading = input.telemetryData.stoppingDirectionHeading;
+        output.useHeading = true;
+        //retrieving desired slow flight speed
```
I'm a bit confused about where we are using this?

Pop0097 commented on Autopilot/PathManager/Src/pathStateClasses.cpp:
```diff
@@ -120,4 +324,360 @@ pathManagerState& fatalFailureMode::getInstance()
+        pathMgr->madeTakeoffPoints = true;
+    }
+
+    if(input.sensorOutput.airspeed>(TakeoffManager::desiredRotationSpeed(pathMgr->isPackage)))
```
space around > pls

Pop0097 commented on Autopilot/Test/Src/PathManager/Test_PathManager_Fsm.cpp:
```diff
@@ -0,0 +1,433 @@
+	EXPECT_EQ(*(pathMan.getCurrentState()), takeoffClimbStage::getInstance());
+}
+
+TEST (PathManagerFSM, TransitionToCT) {
```
TransitionToCoordinatedTurnElevation

This will make the test names more descriptive!

Pop0097 commented on Autopilot/Test/Src/PathManager/Test_PathManager_Fsm.cpp:
```diff
@@ -0,0 +1,433 @@
+/*
+* Author: Dhruv Rawat
+*/
```
Feel free to add your name here!

Pop0097 commented on Autopilot/Test/Src/PathManager/Test_PathManager_Fsm.cpp:
```diff
@@ -0,0 +1,433 @@
+* Author: Dhruv Rawat
+*/
+
+#include "fff.h"
+#include <gtest/gtest.h>
```
Would it make sense to add tests to see the changes in the enum? Like transition --> slope, etc.?

NixonCTChan replied:
it would require some fiddling with the sensor data, but yeah I'll make some.

NixonCTChan replied:
hmm I've tried doing that, but the values that determine if the plane is going to switch enums get updated in the execute function. It's gonna require a lot of sensor measurements and other variables in order to make the state machine think that it should be switching enums.

Pop0097 commented on Autopilot/Test/Src/PathManager/Test_PathManager_Fsm.cpp:
```diff
@@ -0,0 +1,433 @@
+ _FlightStage Enum Testing
+ ***********************************************************************************************************/
+
+TEST (PathManagerFSM, ResetVariablesToLandingTransition) {
+	/***********************SETUP***********************/
+
+	pathManager pathMan;
+	pathMan.stage = TRANSITION;
+
+	/********************DEPENDENCIES*******************/
+	/********************STEPTHROUGH********************/
+
+	pathMan.setState(resetVariables::getInstance());
+	pathMan.execute();
+
+	/**********************ASSERTS**********************/
+
+	EXPECT_EQ(*(pathMan.getCurrentState()), landingTransitionStage::getInstance());
+}
```
Great job with the unit tests! I especially love how readable they are!

Pop0097 commented on Autopilot/PathManager/Src/landingTakeoffManager.cpp:
```diff
@@ -0,0 +1,204 @@
+    Vector3D projectedPoint;
+
+    //calculating the vectors
+    Vector3D vectorAI(iPoint.x - aPoint.x, iPoint.y - aPoint.y, iPoint.z - aPoint.z);
+    Vector3D vectorAS(sPoint.x - aPoint.x, sPoint.y - aPoint.y, sPoint.z - aPoint.z);
+    Vector3D vectorAC(cPoint.x - aPoint.x, cPoint.y - aPoint.y, cPoint.z - aPoint.z);
```
Maybe add comments saying "Aiming --> Intersection"? 

Comment format
// (starting point) --> (ending Point) 

if you think it isn't needed, feel free to ignore!

NixonCTChan replied:
sure i'll add them, improves readability too.

Pop0097 commented on Autopilot/PathManager/Src/landingTakeoffManager.cpp:
```diff
@@ -0,0 +1,204 @@
+    double metersPerDegLon = 40075000.0 * cos(input.stoppingLatitude*PI/180)/360.0; //111251.6316
+    double stoppingDistLon = stoppingDistX / metersPerDegLon; //6.355922792e-5
+    double stoppingDistLat = stoppingDistY / METERS_PER_DEG_LAT; //6.352019235e-5
+
+ 
```
remove two blank lines here pls

Pop0097 commented on Autopilot/PathManager/Src/landingTakeoffManager.cpp:
```diff
@@ -0,0 +1,204 @@
+
+_LandingPath LandingManager::createSlopeWaypoints(Telemetry_PIGO_t input, double currentAltitude)
+{
+    //created comments for testing
```
May want to remove these comments before merging

Pop0097 commented on Autopilot/PathManager/Src/landingTakeoffManager.cpp:
```diff
@@ -0,0 +1,204 @@
+    double stoppingDistY = cos(radianDirection) * DISTANCE_OF_LANDING; //5root2
+
+    //converting into x and y components in lat and lon
+    double metersPerDegLon = 40075000.0 * cos(input.stoppingLatitude*PI/180)/360.0; //111251.6316
```
would it make sense to make 40075000.0 a constexpr?

NixonCTChan replied:
sure I can do that!

Pop0097 reviewed (commented):
Damn! Really good work!

Pop0097 commented on Autopilot/PathManager/Src/pathStateClasses.cpp:
```diff
@@ -120,4 +324,377 @@ pathManagerState& fatalFailureMode::getInstance()
+        output.desiredAirspeed = LandingManager::approachSpeed(pathMgr->isPackage);
+    }
+    
+    if(!landingTransitionStage::waypointStatus == WAYPOINT_SUCCESS)
+    {
+        pathMgr->isError = true;
+    }
```
Should be `landingTransitionStage::waypointStatus != WAYPOINT_SUCCESS` since it's an integer and not a boolean.

Pop0097 replied:
I think you may have done this in other places too

NixonCTChan replied:
oops, messed up the precedences. Will fix

Pop0097 reviewed (commented):
Just one little thing from me. @Fberbari what say you?

Pop0097 commented on Autopilot/PathManager/Inc/pathStateClasses.hpp:
```diff
@@ -18,72 +28,232 @@ class commsWithAttitude : public pathManagerState
+        static _WaypointManager_Data_Out waypointOutput;
+        static _PathData * pathArray[3]; //used to translate loaded in path to something the waypoint manager can take as a parameter
+        static _PathData * currentLocation;
+        double differenceInHeading1;
+        double differenceInHeading2;
```
Another thing actually, I don't think we even need these to be member variables for this class. 

Instead, you can declare them in the ::execute() method for this class. That way we aren't populating this class structure with stuff that aren't really being cached or accessed elsewhere.

NixonCTChan replied:
true, I'll make those

Fberbari commented on Autopilot/.vscode/settings.json:
```diff
@@ -0,0 +1,5 @@
+{
```
Should get rid of this file

Fberbari commented on Autopilot/AttitudeManager/Inc/GetFromPathManager.hpp:
```diff
@@ -2,14 +2,15 @@
  * Author: Anthony Berbari
  */
-
+#include "pathDatatypes.hpp"
```
This should be inside the #ifdef  guard. Reason being that if this file is included recursively, the guard prevents its contents from showing up twice. In practice, I've never seen a case where things are not encapsulated inside the guard

Fberbari commented on Autopilot/PathManager/Inc/landingTakeoffManager.hpp:
```diff
@@ -0,0 +1,163 @@
+struct _LandingPath
+{
+    _PathData stoppingPoint;
+    _PathData aimingPoint;
```
Comments on these things pls (and even a link to your confluence entry). Otherwise it's really intimidating to understand.

Fberbari commented on Autopilot/PathManager/Inc/landingTakeoffManager.hpp:
```diff
@@ -0,0 +1,163 @@
+
+struct _LandingTakeoffInput
+{
+    Telemetry_PIGO_t telemetryData;
```
Note to myself. These things should be passed by reference and should be read only.

Fberbari commented on Autopilot/PathManager/Inc/landingTakeoffManager.hpp:
```diff
@@ -0,0 +1,163 @@
+{
+    double desiredHeading;
+    double desiredTrack;
+    double useHeading; //pay attention
```
THis comment doesen't mean anything atm.

Fberbari commented on Autopilot/PathManager/Src/pathStateClasses.cpp:
```diff
@@ -90,16 +231,79 @@ pathManagerState& cruisingState::getInstance()
+            LandingManager::translateLTSFCommandsToCoordTurns(landingFlareStage::getControlOutput(), sensFusionOutput, imudata, &turnInput, &altAirspeedInput);
+            break;
+        case DECRAB:
+            LandingManager::translateLTSFCommandsToCoordTurns(landingDecrabStage::getControlOutput(), sensFusionOutput, imudata, &turnInput, &altAirspeedInput);
```
I think the only thing changing here is where your'e getting the control input from. So that should be the only thing in the switch statement, the common stuff should just be a single line below the switch statement.

NixonCTChan replied:
the cruising state outputs are accessed completely differently though, that's why I just have the accessing functions in each case statement. I'm not sure how else I can implement this more simply.

Fberbari replied:
What I mean is, you can make it so that rather than having all this in each statement

 LandingManager::translateLTSFCommandsToCoordTurns(landingDecrabStage::getControlOutput(), sensFusionOutput, imudata, &turnInput, &altAirspeedInput);
 
 you can have:
 
 ControlInput_t controlInput =  LandingDecrabStage::getControlOutput();
 
 (with the variabe being asigned whatever it needs to be, then at the bottom, you can have a single call to the function with:
 
  LandingManager::translateLTSFCommandsToCoordTurns(controlInput , sensFusionOutput, imudata, &turnInput, &altAirspeedInput);
  
  
 The 2 reasons I note this:
  
  I'm sure you've heard it said that repetition in software is bad. This is a case of that, if for whatever reason, you have to change the names of the other args or of the function or whatever, you'l have to make that change in each case with your current implementation, which is prone to mistakes and is a bit of a pain.
  
  Second reason is that this current implementation forces readers to read every line, playing "find the differences" to figure out what your'e trying to do. That's much more tough than being able to just scan it over.
  
  
  
  
  Is it the biggest issue in the world ? no. But if you get in the habbit of doing little things like this, it'l slowly make you a better developer.

NixonCTChan replied:
ok gotcha, I'll make those changes

Fberbari commented on Autopilot/PathManager/Src/pathStateClasses.cpp:
```diff
@@ -90,16 +231,79 @@ pathManagerState& cruisingState::getInstance()
+            LandingManager::translateLTSFCommandsToCoordTurns(takeoffRollStage::getControlOutput(), sensFusionOutput, imudata, &turnInput, &altAirspeedInput);
+            break;
+        case CLIMB:
+            LandingManager::translateLTSFCommandsToCoordTurns(takeoffClimbStage::getControlOutput(), sensFusionOutput, imudata, &turnInput, &altAirspeedInput);
```
Climb is being taken care of by the landing manager ? That's very unintuitive

NixonCTChan replied:
The function for translating both landing and takeoff outputs to coord turn commands lives in the landing manager class. Should I make it a stand alone function?

Fberbari replied:
I'll take a closer look and lyk, this may be the best solution, but it feels weird

Fberbari replied:
Ok I see. Really it's the landingTakeoff manager, no problem. I think changing the name to that would make things a bit clearer. THoughts ?

Fberbari commented on Autopilot/PathManager/Src/pathStateClasses.cpp:
```diff
@@ -120,4 +324,360 @@ pathManagerState& fatalFailureMode::getInstance()
+void landingTransitionStage::execute(pathManager* pathMgr)
+{
+    //load in sensor fusion data and telemtry data into input structure
+    input.telemetryData = *(commsWithTelemetry::GetTelemetryIncomingData());
```
Why are we passing these structs by value ? They're read only. They should be passed as constant references or something C++y of the sort.

Fberbari reviewed (commented):
Ok, ya, this is a lot. Which is my own fault for neglecting to review this fr some time. But there's some really great stuff here, really great, but it's really overwhelming too. 

There's a couple things I'm nervous about, like the complexity of path manager, which seems to be growing at an insane rate.


Regardless, with 33 files to review, I'm getting really confused about what changes are yours and what's in here because of the rebasing issues. 

Prioritize fixing the rebasing crap, and push again and I'll complete a deep review.

Fberbari commented on Autopilot/PathManager/Inc/landingTakeoffManager.hpp:
```diff
@@ -0,0 +1,167 @@
+{
+    _PathData stoppingPoint;
+    _PathData aimingPoint;
+    _PathData intersectionPoint;
```
hate to be nitpicky, but what I really meant was explaining things like what the intersection point is in a few words would make things much more understandable.

Fberbari commented on Autopilot/PathManager/Inc/landingTakeoffManager.hpp:
```diff
@@ -0,0 +1,167 @@
+
+            @return double - this function will return the desired rotation speed for the aircraft
+        */
+        static double desiredRotationSpeed(const bool * ifPackage);
```
As a general rule. Individual built in variables, like bools, ints, and doubles, should be passed by value. Structs should be passed by reference.

NixonCTChan replied:
oh yeah ok, that makes sense. Will fix

Fberbari commented on Autopilot/PathManager/Inc/trackFinder.hpp:
```diff
@@ -1,7 +1,9 @@
 #pragma once
```
THis file does not need to exist. Track is just pulled from the sensor fusion output. The guts of this file should live inside SensorFusion.cpp

NixonCTChan replied:
I don't think sensorFusion has those changes yet though?

Fberbari commented on Autopilot/PathManager/Src/landingTakeoffManager.cpp:
```diff
@@ -0,0 +1,202 @@
+    return altitude;
+}
+
+double LandingManager::approachSpeed(const bool *ifPackage)
```
noted before, but emphasizing here. Reason to pass by reference is to save copying space/time. THat's good for custom structs, but since a bool and a pointer to a bool take up the same space, there's no need to pass a bool by reference.

Fberbari commented on Autopilot/PathManager/Src/landingTakeoffManager.cpp:
```diff
@@ -0,0 +1,202 @@
+    }
+}
+
+_LandingPath LandingManager::createSlopeWaypoints(const Telemetry_PIGO_t * input, const float * currentAltitude)
```
THere is a c++y way of passing these variables by reference. What you've got here is the old C way, which I started building into our code cuz I didn't know any better. Here's the right way to do it in C++.
https://www.ibm.com/docs/en/zos/2.2.0?topic=calls-pass-by-reference-c-only

The advantage to this is you don't ever have to worry about pointer de referencing or people passing in trash pointers ever again.

Fberbari commented on Autopilot/Src/SensorFusion.cpp:
```diff
@@ -6,7 +6,13 @@
+typedef struct {
+    float roll, pitch, yaw; //rad
+    float rollRate, pitchRate, yawRate; //rad/s
+    float Airspeed; //m/s
```
track was added into here some time ago. THis is rebasing problem shit. Uh Oh. We're gonna have a hell of a time merging this LOOOL but we'll cross that bridge when we come to it.

Don't bother doing anything about it now though.

Fberbari reviewed (changes requested):
Ok, done. I think we're almost good to merge. I didn't detect any major problems here, so once you fix up the minor stuff I noted, I'll approve.

NixonCTChan commented:
- landing and takeoff finished!

Fberbari commented:
Closing cuz these changes went into another, cleaner PR that's now merged. This PR faced too many merged conflicts.
